### PR TITLE
[ODPO] Fix global step for consistent checkpointing with global updates

### LINKED ISSUE #1889: [Bug][OnlineDPO] Far too many checkpoints are saved
There is a bug with the `OnlineDPOTrainer` and using `--save_strategy=steps` , I trained a model with `--save_steps=200` , for the first 1800 steps the checkpointing was correct. After that it saved a checkpoint every 2 steps, causing me to write 12TB of data to disc on the HF cluster.

Command to replicate (on a 8 GPU node)
```
accelerate launch --config_file examples/accelerate_configs/deepspeed_zero3.yaml \
    examples/scripts/online_dpo.py \
    --dataset_name=trl-internal-testing/tldr-preference-sft-trl-style \
    --learning_rate=3e-6 \
    --output_dir=data/online_dpo/EleutherAI_pythia-1b \
    --per_device_train_batch_size=16 \
    --gradient_accumulation_steps=4 \
    --total_episodes=1000000 \
    --local_rollout_forward_batch_size=8 \
    --model_name_or_path=cleanrl/EleutherAI_pythia-1b-deduped__sft__tldr \
    --reward_model_path=cleanrl/EleutherAI_pythia-1b-deduped__reward__tldr \
    --non_eos_penalty \
    --stop_token=eos \
    --response_length=53 \
    --save_strategy=steps \
    --save_steps=200 \
    --num_epochs=1 \
    --num_total_batches=1 \
    --beta=0.1 \
    --push_to_hub

```

### PR DESCRIPTION
**Pull request by lewtun**
This PR aligns the `global_step` to increment in the same way as the outer update loop. This ensures that the intermediate checkpoints align with the value set in `save_steps`.

The main impact of this change is that it affects the LR scheduler since the global step in incremented differently. See e.g. these two WandB runs:

* Without fix: https://wandb.ai/huggingface/huggingface/runs/t1v4qfj1?nw=nwuserlewtun
* With fix: https://wandb.ai/huggingface/huggingface/runs/sfebrzr6/overview

This subsequently affects the training dynamics, but I am unsure if the original implementation was actually correct or not wrt the global step.

Screenshot 2024-08-20 at 15 04 13

Closes #1889

### CODE REVIEW

File: trl/trainer/online_dpo_trainer.py [hunk 1]
```diff
@@ -505,11 +503,11 @@ def repeat_generator():
                 metrics["lr"] = self.lr_scheduler.get_last_lr()[0]
                 metrics["episode"] = self.state.episode
                 self.state.epoch = self.state.episode / self.train_dataset_len  # used by self.log
-                self.state.global_step += 1
```
Comment (lewtun) [inline trl/trainer/online_dpo_trainer.py:508]: Moved this to occur after the LR scheduler update, similar to how `transformers.Trainer` does it.

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
also added to #1943
